feat/FECA-51 orders page api

YassenAhmed-gif opened the pull request:
<!--
  🛠️ For Work In Progress, please use Draft PRs:
  https://github.blog/2019-02-14-introducing-draft-pull-requests/

  ✅ Before submitting:
  - Follow your project's contributing guide
  - Ensure code compiles and passes all tests
  - Use descriptive commit messages
  - Keep PRs small and focused
-->

## 📌 What type of PR is this? (check all applicable)

- [ ] ✨ Feature
- [ ] 🐛 Bug Fix
- [ ] 🧹 Refactor
- [ ] 🚀 Performance Optimization
- [ ] 🧾 Documentation Update
- [ ] ♿ Accessibility Fix
- [ ] 🔧 CI/CD or Config Change

---

## 📝 Description

_Describe the goal of this PR and what was changed._

---

## 🧩 Related Issues / Tickets

- Closes #
- Fixes #
- Related to #

---

## 🧪 QA Instructions

_How to test the changes manually?_

- [ ] Run the app on Android
- [ ] Run the app on iOS
- [ ] Check specific feature/page:
  - [ ] ...
- [ ] Verify responsiveness on different screen sizes

---

## 📸 Screenshots or Recordings (if UI changes)

_Add screenshots or screen recordings to demonstrate your changes._

---

## ♿ UI Accessibility Checklist (if applicable)

- [ ] Semantic widget structure used (`Semantics`, `Accessible`, etc.)
- [ ] Supports screen readers
- [ ] Good color contrast
- [ ] Tested with TalkBack/VoiceOver

---

## 🧪 Tests

- [ ] Unit tests added
- [ ] Widget tests added
- [ ] Integration tests added
- [ ] No tests needed (explain why):

---

## 🛠 Post-Deployment Tasks (Optional)

_Any migrations, scripts, or setup tasks?_

---

## 🎬 Fun GIF (Optional)

_What GIF best represents this PR?_

![fun gif](https://media.giphy.com/media/xT0xeJpnrWC4XWblEk/giphy.gif)

YassenAhmed-gif committed:
feat: Implement Orders Page

This commit introduces the Orders Page, which displays a user's active and completed orders.

Key changes:
- Added new entities: `OrderItems`, `OrdersEntity`, `MetadataEntity`, and `OrdersResponseEntity` to represent order data.
- Created `OrdersScreen` to display orders in two tabs: "Active" and "Completed".
- Implemented `OrderCard` widget to display individual order item details.
- Added `OrdersViewModel` to manage the state for fetching and displaying orders, along with `OrdersEvent` and `OrdersState`.
- Created `OrdersUseCase` with mock data for fetching orders.
- Added new translation keys for "my_orders", "active", "completed", "no_orders", "no_active_orders", "no_completed_orders", "delivered_on", "order_number", "reorder", and "track_order" in English and Arabic.
- Updated `AppRoutes` to include `ordersRoute`.
- Modified `RouteGenerator` to handle the new `ordersRoute`.
- Added new size constants in `AppSizes` for card dimensions.
- Updated `ProductDetailsModel` to include `priceAfterDiscount`, `discount`, and `imageCover`.
- Fixed `AlignmentGeometry` to `Alignment` in `ProductCard`.
- Updated `SignInScreen` to use translated locale keys.
- Added `ordersRoute` as the initial route in `main.dart` for development purposes.
- Added unit tests for `OrdersViewModel` in `orders_view_model_test.dart`.

YassenAhmed-gif committed:
Feat: Implement Orders API and Data Handling

This commit introduces the necessary DTOs, mappers, service, and data source for fetching and handling order data from the API.

Key changes:
- Added new DTOs for `OrdersDto`, `OrderItemsDto`, `ProductDto`, `OrdersResponseDto`, and `MetaDataDto` to represent the API response structure. These include JSON serialization/deserialization logic.
- Implemented corresponding mappers (`ProductMapper`, `OrderItemsMapper`, `OrdersMapper`, `OrderResponseMapper`, `MetadataMapper`) to convert DTOs to domain entities.
- Created `OrdersService` with a `getOrders` method to fetch order data from the `ApiConstants.orders` endpoint.
- Implemented `OrdersRemoteDataSourceImpl` to utilize the `OrdersService` and map the API response to `OrdersResponseEntity`.
- Updated `OrderItems` entity to `OrderItemsEntity` and adjusted the `OrdersEntity` to use `OrderItemsEntity`.
- Added `orders` to `ApiConstants`.

YassenAhmed-gif committed:
feat: Add OrdersRemoteDataSource and OrdersRepoImpl

This commit introduces the initial data layer components for the Orders feature.

Key changes:
- Created `OrdersRemoteDataSource.dart` defining an abstract interface with a `getAllOrders()` method.
- Added `OrdersRepoImpl.dart` which implements the `OrdersRepo` interface and depends on `OrdersRemoteDataSource` to fetch orders.

YassenAhmed-gif committed:
Refactor: Update Orders feature components and tests

This commit introduces several changes to the Orders feature, including refactoring data sources, repositories, use cases, and updating related tests.

Key changes:
- Renamed `OrderItems` to `OrderItemsEntity` in `orders_view_model_test.dart` and `orders_screen.dart` for better clarity.
- Updated `OrdersRemoteDataSourceImpl` and `OrdersRepoImpl` to use `@Injectable(as: ...)` for proper dependency injection.
- Refactored `OrdersUseCase` to use `OrdersRepo` for fetching orders, removing hardcoded data.
- Introduced `OrdersRepo` abstract interface.
- Ensured `OrdersRepoImpl` implements the new `OrdersRepo` interface.

YassenAhmed-gif committed:
Test: Add comprehensive tests for Orders feature and refactor related components

This commit introduces a full suite of tests for the Orders feature, covering the ViewModel, Remote Data Source, Repository, and Use Case. It also includes refactoring of the Orders feature components and related DTOs.

Key changes:
- Added `orders_view_model_test.dart` to verify state changes in `OrdersViewModel` for success, failure, and empty order scenarios. Ensured correct handling of loading states and error messages.
- Added `orders_remote_data_source_impl_test.dart` to test `OrdersRemoteDataSourceImpl`, covering successful API calls, various error scenarios (network, mapping, invalid response, timeout), and correct data mapping.
- Added `orders_repo_impl_test.dart` to test `OrdersRepoImpl`, ensuring correct data propagation from the data source, handling of success and error results, empty order lists, multiple orders, specific failure types, and null order data.
- Added `orders_use_case_test.dart` to test `OrdersUseCase`, verifying interaction with the repository, handling of success and error results, empty and complex order data, and various failure scenarios.
- Refactored `OrdersRemoteDataSourceImpl` to use try-catch for error handling and return `ApiResult` directly.
- Updated `ProductDto` to include `Id` (mapped from `_id`) and `v` fields, and removed the `discount` field.
- Updated `ProductMapper` to use the new `Id` field from `ProductDto`.
- Updated `OrdersState` in `orders_state.dart` to reorder fields and improve `copyWith` logic for clearing failure and orders.
- Updated `OrdersViewModel` to align with `OrdersState` changes.
- In `OrdersScreen`, added a `Scaffold` with `CustomBackButton` and `ProductsShimmer` for the loading state. Ensured `imgCover` in `OrderCard` handles null values.
- In `OrderCard`, ensured `imgCover` in `CachedNetworkImage` handles null values.
- Corrected state assertions in `orders_view_model_test.dart` for loading, success, and failure states to be more precise.

Eng-M0ataz commented on lib/Feature/ordersPage/api/client/orders_service.dart:
```diff
@@ -0,0 +1,19 @@
+
+@RestApi()
+@injectable
+abstract class OrdersService {
```
ordersApiService

Eng-M0ataz commented on lib/Feature/ordersPage/api/client/orders_service.dart:
```diff
@@ -0,0 +1,19 @@
+import 'package:injectable/injectable.dart';
+import 'package:retrofit/error_logger.dart';
+import 'package:retrofit/http.dart';
+
```
remove this space

Eng-M0ataz commented on lib/Feature/ordersPage/api/client/orders_service.dart:
```diff
@@ -0,0 +1,19 @@
+
```
remove this space

Eng-M0ataz commented on lib/Feature/ordersPage/api/dataSources/orders_remote_data_source_impl.dart:
```diff
@@ -0,0 +1,31 @@
+
```
remove this space

Eng-M0ataz commented on lib/Feature/ordersPage/api/dataSources/orders_remote_data_source_impl.dart:
use execute api function

Eng-M0ataz commented on lib/Feature/ordersPage/api/mappers/metadata_mapper.dart:
```diff
@@ -0,0 +1,14 @@
+
```
remove space

Eng-M0ataz commented on lib/Feature/ordersPage/api/mappers/metadata_mapper.dart:
```diff
@@ -0,0 +1,14 @@
+extension MetadataMapper on MetaDataDto {
+  MetadataEntity toEntity() {
+    return MetadataEntity(
+      currentPage: currentPage,
```
handle null value

Eng-M0ataz commented on lib/Feature/ordersPage/api/mappers/metadata_mapper.dart:
```diff
@@ -0,0 +1,14 @@
+  MetadataEntity toEntity() {
+    return MetadataEntity(
+      currentPage: currentPage,
+      limit: limit,
```
handle null value of all parm

Eng-M0ataz commented on lib/Feature/ordersPage/api/mappers/response/order_response_mapper.dart:
```diff
@@ -0,0 +1,15 @@
+  OrdersResponseEntity toEntity() {
+    return OrdersResponseEntity(
+      message: message ?? '',
+      metaData: metadata?.toEntity(),
```
handle null value

Eng-M0ataz commented on lib/Feature/ordersPage/api/models/metadata_dto.dart:
```diff
@@ -0,0 +1,32 @@
+import 'package:json_annotation/json_annotation.dart';
+
```
remove space

Eng-M0ataz commented on lib/Feature/ordersPage/api/models/order_items_dto.dart:
```diff
@@ -0,0 +1,31 @@
+import 'package:flower_e_commerce_app/Feature/ordersPage/api/models/product_dto.dart';
+import 'package:json_annotation/json_annotation.dart';
+
```
remove space

Eng-M0ataz commented on lib/Feature/ordersPage/api/models/orders_dto.dart:
```diff
@@ -0,0 +1,52 @@
+import 'package:flower_e_commerce_app/Feature/ordersPage/api/models/order_items_dto.dart';
+import 'package:json_annotation/json_annotation.dart';
+
```
remove space

Eng-M0ataz commented on lib/Feature/ordersPage/api/models/product_dto.dart:
```diff
@@ -0,0 +1,75 @@
+import 'package:json_annotation/json_annotation.dart';
+
```
remove space

Eng-M0ataz commented on lib/Feature/ordersPage/api/models/response/orders_response_dto.dart:
```diff
@@ -0,0 +1,29 @@
+import 'package:flower_e_commerce_app/Feature/ordersPage/api/models/metadata_dto.dart';
+import 'package:flower_e_commerce_app/Feature/ordersPage/api/models/orders_dto.dart';
+import 'package:json_annotation/json_annotation.dart';
+
```
remove space

Eng-M0ataz commented on lib/Feature/ordersPage/data/repositories/orders_repo_impl.dart:
```diff
@@ -0,0 +1,17 @@
+
```
remove space

Eng-M0ataz commented on lib/Feature/ordersPage/domain/entities/order_entity.dart:
```diff
@@ -0,0 +1,31 @@
+import 'order_items.dart';
+
```
remove space

Eng-M0ataz commented on lib/Feature/ordersPage/presentation/pages/orders_screen.dart:
refacor all the screen 200 line is to much for this small ui and i cant read any thing from you code refacor this screen well and extract widgets as custom widgets and put them in sepertad files

YassenAhmed-gif replied:
I made a Widget 'ShowOrders' to split the code of presenting the orders after ApiSuccessResult
from rest of the screen

Eng-M0ataz commented on lib/Feature/ordersPage/presentation/viewModels/orders_state.dart:
```diff
@@ -0,0 +1,42 @@
+    return OrdersState(
+      isLoading: isLoading ?? this.isLoading,
+      isSuccess: isSuccess ?? this.isSuccess,
+      orderFailure: clearFailure ? null : (orderFailure ?? this.orderFailure),
```
what is this for ?

YassenAhmed-gif replied:
i used them in the process of resolving some bugs, and i forgot to remove them
the will be removed

Eng-M0ataz commented on lib/Feature/ordersPage/presentation/widgets/order_card.dart:
```diff
@@ -0,0 +1,101 @@
+import 'package:flower_e_commerce_app/core/Widgets/custom_elevated_button.dart';
+import 'package:flower_e_commerce_app/core/localization/locale_keys.g.dart';
+import 'package:flutter/material.dart';
+
```
remove space

Eng-M0ataz commented on lib/core/localization/locale_keys.g.dart:
dont upload this

YassenAhmed-gif replied:
i will remove it from the PR👍

Eng-M0ataz commented on lib/core/models/product_details_model.dart:
why you make changes is this file ?

YassenAhmed-gif replied:
I used it as a product entity for the data that is coming from the response
and I needed to store the imageCover & priceAfterDiscount course those will be presented in the card discount I did not found any use for it after I uploaded the PR, so I well remove it.

YassenAhmed-gif committed:
Refactor: Enhance Orders Feature

This commit introduces several improvements to the Orders feature, including UI refactoring, code cleanup, and enhanced error handling.

Key changes:
- Refactored `OrdersScreen` by extracting the orders display logic into a new widget `ShowOrders`. This improves code organization and readability.
- Removed unnecessary imports and newlines across various files in the Orders feature, including DTOs, mappers, services, and repositories.
- Renamed `OrdersService` to `OrdersApiService` for better clarity and consistency.
- Updated `OrdersRemoteDataSourceImpl` to utilize the `executeApi` helper function for streamlined API call handling and error management.
- Improved null safety in `MetadataMapper` and `OrderResponseMapper` by providing default values for nullable fields, ensuring more robust data mapping.
- Adjusted `OrdersState` by removing the `clearFailure` and `clearOrders` parameters from the `copyWith` method as they are no longer needed.
- Removed the unused `discount` field from `ProductDetailsModel`.

YassenAhmed-gif committed:
refactor: update imports, constants, and test mocks for orders feature

This commit refactors several parts of the codebase, primarily related to the orders feature and constant management.

Key changes include:
- Deleted the old `api_constants.dart` file and consolidated the `orders` endpoint into the new `Constantts/api_constants.dart` file.
- Corrected import paths in multiple files to point to the new constants location (`Constantts/`) and updated `execute_api.dart` path.
- Renamed `MockOrdersService` to `MockOrdersApiService` in `orders_remote_data_source_impl_test.dart` for better consistency with the service it mocks.

YassenAhmed-gif committed:
Merge remote-tracking branch 'origin/development' into feat/FECA-51-Orders-Page-API

# Conflicts:
#	assets/translations/ar.json
#	lib/core/Config/Routing/route_generator.dart
#	lib/core/utils/Constantts/api_constants.dart
#	lib/core/utils/Constantts/app_routes.dart
#	lib/main.dart
#	pubspec.lock
#	pubspec.yaml
